### PR TITLE
Load Shared Libraries Dynamically

### PR DESCRIPTION
**Pull request by dyastremsky**
Allow user to load TensorRT shared libraries via the backend configuration.

Related TensorRT backend ticket: https://github.com/triton-inference-server/tensorrt_backend/pull/25

### CODE REVIEW

**Comment by dyastremsky** [comment]
@CoderHam Could you look to see if there's something off? The TensorRT team confirmed the clip plugin is maintained and regularly tested. Perhaps there's a mismatch somewhere in terms of TensorRT version, inputs, outputs, or something minor.

File: qa/common/gen_qa_trt_plugin_models.py [hunk 1]
```diff
@@ -102,9 +101,22 @@ def get_trt_plugin(plugin_name):
...
+            max_clip = trt.PluginField("clipMax", np.array([0.5],\
+                dtype=np.float32), trt.PluginFieldType.FLOAT32)
+            field_collection = trt.PluginFieldCollection([min_clip, max_clip])
+            break
+
+    if (field_collection is not None):
+        plugin = plugin_creator.create_plugin(name=plugin_name,
+                                              field_collection=field_collection)
+    else:
+        raise RuntimeError("Plugin not found: " + plugin_name)
```
Comment (CoderHam) [inline qa/common/gen_qa_trt_plugin_models.py:118]: Why not raise the actual error instead?

**Review by CoderHam — CHANGES_REQUESTED** [review]

**Comment by dyastremsky** [comment]
> @CoderHam Could you look to see if there's something off? The TensorRT team confirmed the clip plugin is maintained and regularly tested. Perhaps there's a mismatch somewhere in terms of TensorRT version, inputs, outputs, or something minor.

Update on this comment: TRT 14.0 does not have this error, set to release around the end of May. In the meanwhile, I patched the few lines of code in their sample plugin to generate the plugin and model for our CI locally. The tests now pass. I attached the model, plugin, and issue link (with the patch) to our internal Jira ticket.

**Comment by dyastremsky** [comment]
> @CoderHam Could you look to see if there's something off? The TensorRT team confirmed the clip plugin is maintained and regularly tested. Perhaps there's a mismatch somewhere in terms of TensorRT version, inputs, outputs, or something minor.

Update on this comment: An upcoming release of TensorRT should not have this issue. In the meanwhile, I patched the few lines of code in their sample plugin to generate the plugin and model for our CI locally. The tests now pass. I attached the model, plugin, and issue link (with the patch) to our internal Jira ticket.

File: qa/L0_trt_plugin/test.sh [hunk 2]
```diff
@@ -63,14 +153,15 @@ if [ "$SERVER_PID" == "0" ]; then
     exit 1
 fi
 
+rm -f $CLIENT_LOG
 set +e
-python $PLUGIN_TEST >$CLIENT_LOG 2>&1
+python $PLUGIN_TEST PluginModelTest.test_raw_fff_clip
```
Comment (tanmayv25) [inline qa/L0_trt_plugin/test.sh:158]: You seem to be missing `>>$CLIENT_LOG 2>&1` here.

File: qa/L0_trt_plugin/test.sh [hunk 3]
```diff
@@ -43,18 +43,108 @@ export CUDA_VISIBLE_DEVICES=0
...
 RET=0
+rm -f ./*.log
+
+LOG_IDX=0
+
+## Default Plugin Tests
+
+## Create model folder with default plugin models
+rm -fr models && mkdir -p models
+find $DATADIR/qa_trt_plugin_model_repository/ -mindepth 1 -maxdepth 1 ! -iname '*clipplugin*' -exec cp -r {} models \;
```
Comment (tanmayv25) [inline qa/L0_trt_plugin/test.sh:72]: Can you add a test case where we have multiple models in the model repo and test that plugins are loaded only once?

**Review by tanmayv25 — CHANGES_REQUESTED** [review]

File: qa/L0_trt_plugin/test.sh (hunk 2 above)
Comment (dyastremsky) [reply qa/L0_trt_plugin/test.sh]: Fixed, thanks.

**Comment by dyastremsky** [comment]
@tanmayv25 I have changed requested from you and Hemant. Do you know whether there's a way to merge despite Hemant's request for changes once you approve? Or do I need to cherry pick?

**Review by tanmayv25 — APPROVED** [review]

**Comment by tanmayv25** [comment]
@dyastremsky I think the best way forward would be to create a new PR and merge the changes. Hemat's requested change will not allow you to merge the PR.

**Comment by dyastremsky** [comment]
> @dyastremsky I think the best way forward would be to create a new PR and merge the changes. Hemat's requested change will not allow you to merge the PR.

Thanks, @tanmayv25! I've copied the commits to #4792 to allow for merging.

**Comment by dyastremsky** [comment]
> @dyastremsky I think the best way forward would be to create a new PR and merge the changes. Hemat's requested change will not allow you to merge the PR.

Got it, thanks for reviewing. I've created #4793 with the cherrypicked changes for approval so that we can get this merged.

**Comment by tanmayv25** [comment]
Closing the PR as the changes have been merged in PR #4793 .

**Pull request closed without merging**